Yoshi/QAKit/EnvironmentMenu/YoshiEncodableEnvironmentManager.swift

```diff
@@ -0,0 +1,70 @@
+//
+//  YoshiEncodableEnvironmentManager.swift
+//  Yoshi
+//
+//  Created by Kanglei Fang on 25/09/2017.
+//  Copyright © 2017 Prolific Interactive. All rights reserved.
+//
+
+import Foundation
+
+/// Environment selection event.
+public typealias EnvironmentChangeEvent = (Environment) -> Void
+
+/// A YoshiEnvironmentManager that persisit user's environment selection using NSUserDefaults.
+open class YoshiEncodableEnvironmentManager: YoshiEnvironmentManager {
+    
+    public var currentEnvironment: Environment {
+        didSet {
+            archive(environment: currentEnvironment)
+            onEnvironmentChange?(currentEnvironment)
+        }
+    }
+    
+    public private(set) var environments: [Environment]
+    
+    private static let YoshiEnvironemntKey = "YoshiEnvironment"
+    
+    private static let encoder = JSONEncoder()
+    private static let decoder = JSONDecoder()
+    
+    private var onEnvironmentChange: EnvironmentChangeEvent?
+    
+    /// Initialize the environement manager with the given environments.
+    /// The manger will retrieve the archaved environment selection from NSUserDefaults if any.
+    /// Othereise the first environment will be used as the current environment.
+    ///
+    /// - Parameters:
+    ///   - environments: Available Environments.
+    ///   - onEnvironmentChange: Callback when environment is changed.
+    ///                          Notice that the callback will be invoked when the manager is initialized.
+    public init(environments: [Environment], onEnvironmentChange: EnvironmentChangeEvent?) {
+        self.environments = environments
+        self.onEnvironmentChange = onEnvironmentChange
+        if let archivedEnvironment = YoshiEncodableEnvironmentManager.archivedEnvironment {
+            currentEnvironment = archivedEnvironment
```

Comment:
What if the `archivedEnvironment` is not in the list of environments passed in?

Reply:
See below

Reply:
I still think you should check if the list of environments passed in contain the `archivedEnvironment`. There could be a number of mismatched environment issues otherwise.